cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -1127,6 +1127,64 @@ class ZeroCopyCast : public CastKernelBase {
   }
 };
 
+class ExtensionCastKernel : public CastKernelBase {
+ public:
+  static Status Make(const DataType& in_type, std::shared_ptr<DataType> out_type,
+                     const CastOptions& options,
+                     std::unique_ptr<CastKernelBase>* kernel) {
+    if (in_type.id() != Type::EXTENSION) {
+      return Status::TypeError("Not an extension type");
+    }
+    const auto storage_type = checked_cast<const ExtensionType&>(in_type).storage_type();
+
+    std::unique_ptr<UnaryKernel> storage_caster;
+    RETURN_NOT_OK(GetCastFunction(*storage_type, out_type, options, &storage_caster));
+    kernel->reset(
+        new ExtensionCastKernel(std::move(storage_caster), std::move(out_type)));
+
+    return Status::OK();
+  }
+
+  Status Init(const DataType& in_type) override {
+    auto& type = checked_cast<const ExtensionType&>(in_type);
+    storage_type_ = type.storage_type();
+    extension_name_ = type.extension_name();
+    return Status::OK();
+  }
+
+  Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
+    DCHECK_EQ(input.kind(), Datum::ARRAY);
+    if (input.type()->id() != Type::EXTENSION) {
+      return Status::TypeError("Not an extension type");
+    }
+
+    // validate: type is the same as the type the kernel was constructed with
+    auto& input_type = checked_cast<const ExtensionType&>(*input.type());
+    if (input_type.extension_name() != extension_name_) {
+      return Status::TypeError(
+          "The cast kernel was constructed with a differently named extension type");
```

Comment:
This error message (and the one below for storage_type) can be more helpful:
```suggestion
          "The cast kernel was constructed to cast from the extension type named '", extension_name_, 
          "' but input has named extension type name '", input_type.extension_name(), "'");
```

Reply:
Indeed.

Reply:
please also make the error for storage type verbose